### ISSUE TITLE
NodeSelector for Jupyter UI and notebooks

### ISSUE THREAD

**Issue by achandras** [kubeflow/kubeflow#5235]
/kind question

I am deploying kubeflow 1.0.2, and the default `spawner-ui-config.yaml` didn't have any indication of where to set node selectors.

I would like the notebook server and notebooks to run on a specifically labeled set of nodes, so any ideas on how to accomplish this would be appreciated

**Comment by MartinForReal** [comment]
```
kind: Notebook
apiVersion: kubeflow.org/v1alpha1
metadata:
  labels:
    app: test
  name: test
  namespace: test
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: resource_type
                    operator: In
                    values:
                      - lycc-p40
      containers:
        - image: 'yourimage'
          imagePullPolicy: Always
          name: test
          resources:
            limits:
              cpu: '8'
              memory: 48Gi
              nvidia.com/gpu: 1
            requests:
              cpu: '8'
              memory: 48Gi
          volumeMounts:
            - mountPath: /dev/shm
              name: dshm
      hostIPC: true
      hostPID: true
      imagePullPolicy: Always
      initContainers:
          image: 'yourimage'
          name: init-myservice
      securityContext:
        fsGroup: 2000
      serviceAccountName: default-editor
      nodeSelector:
         test: a
      tolerations:
        - effect: NoSchedule
          key: allow-gpu
          operator: Exists
      ttlSecondsAfterFinished: 300
      volumes:
        - name: prophet-tutorial
          persistentVolumeClaim:
            claimName: tutorial
            readOnly: true
        - emptyDir:
            sizeLimit: 2000Gi
          name: cachevol
        - emptyDir:
            medium: Memory
          name: dshm
```

FYI

**Comment by achandras** [comment]
@MartinForReal the Notebook resource does indeed let me add pod specs. I was wondering if it was possible to add them through the Jupyter Web App UI.

Many thanks for the pointer!

**Comment by MartinForReal** [comment]
I think one possible solution is that you can mount a template file using configmap and change the content of template accordingly

**Comment by achandras** [comment]
@MartinForReal apologies for the delay in responding. Could you explain a bit more about the template file in configmap? Is there some way to configure the notebook controller to use a template for its deployments?

If there is some documentation or example of this, that would be very helpful

**Comment by thesuperzapper** [comment]
@achandras PR #5237 added support for tolerations/affinity in the Notebook UI, but that is only on `master`, not in any release yet.

**Comment by kimwnasptd** [comment]
This feature is implemented in #5237 as @thesuperzapper and is also merged in the `crud-web-apps/jupyter` app.

The latest Jupyter web app's config allows the user to handle tolerations/affinity so I believe we can close this
